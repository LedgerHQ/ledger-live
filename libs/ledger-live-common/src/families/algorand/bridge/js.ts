import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import type { AlgorandTransaction } from "../types";
import { sync, scanAccounts } from "../js-synchronization";
import {
  createTransaction,
  prepareTransaction,
} from "../js-prepareTransaction";
import { estimateMaxSpendable } from "../js-estimateMaxSpendable";
import { getTransactionStatus } from "../js-getTransactionStatus";
import { signOperation } from "../js-signOperation";
import { broadcast } from "../js-broadcast";
import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";

const receive = makeAccountBridgeReceive();

const updateTransaction = (t, patch) => {
  return { ...t, ...patch };
};

const preload = async () => Promise.resolve({});

const hydrate = () => {};

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<AlgorandTransaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  sync,
  receive,
  signOperation,
  broadcast,
  estimateMaxSpendable,
};

export default {
  currencyBridge,
  accountBridge,
};
