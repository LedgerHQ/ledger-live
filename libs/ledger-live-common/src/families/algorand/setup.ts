// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
