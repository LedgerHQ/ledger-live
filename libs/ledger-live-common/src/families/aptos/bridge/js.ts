import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import { getSerializedAddressParameters } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../synchronisation";
import getTransactionStatus from "../getTransactionStatus";
import prepareTransaction from "../prepareTransaction";
import createTransaction from "../createTransaction";
import estimateMaxSpendable from "../estimateMaxSpendable";
import signOperation from "../signOperation";
import broadcast from "../broadcast";

const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};

const updateTransaction = (t: Transaction, patch: Partial<Transaction>): Transaction => ({
  ...t,
  ...patch,
});

const receive = makeAccountBridgeReceive();

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  getSerializedAddressParameters,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
