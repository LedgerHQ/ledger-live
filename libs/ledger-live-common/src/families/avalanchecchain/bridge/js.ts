import type { AccountBridge, CurrencyBridge } from "../../../types";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import { createTransaction, prepareTransaction, updateTransaction } from '../js-transaction';
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from '../js-signOperation';
import broadcast from "../js-broadcast";
import estimateMaxSpendable from '../js-estimateMaxSpendable';

const preload = () => Promise.resolve({});
const hydrate = (): void => { };

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
    preload,
    hydrate,
    scanAccounts,
};
const accountBridge: AccountBridge<Transaction> = {
    estimateMaxSpendable,
    createTransaction,
    updateTransaction,
    getTransactionStatus,
    prepareTransaction,
    sync,
    receive,
    signOperation,
    broadcast,
};
export default {
    currencyBridge,
    accountBridge,
};
