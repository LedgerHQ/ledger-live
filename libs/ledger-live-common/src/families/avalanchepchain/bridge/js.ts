import type { AccountBridge, CurrencyBridge } from "@ledgerhq/types-live";
import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";
import { sync, scanAccounts } from "../js-synchronisation";
import {
  createTransaction,
  prepareTransaction,
  updateTransaction,
} from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import { preload, hydrate } from "../js-preload-data";
import estimateMaxSpendable from "../js-estimateMaxSpendable";

const receive = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  preload,
  hydrate,
  scanAccounts,
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};
export default {
  currencyBridge,
  accountBridge,
};
