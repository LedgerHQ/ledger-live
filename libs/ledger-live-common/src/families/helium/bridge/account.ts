import { makeAccountBridgeReceive, makeSync } from "../../../bridge/jsHelpers";
import { AccountBridge } from "@ledgerhq/types-live";
import broadcast from "../js-broadcast";
import estimateMaxSpendable from "../js-estimateMaxSpendable";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import {
  createTransaction,
  prepareTransaction,
  updateTransaction,
} from "../js-transaction";
import { Transaction } from "../types";
import { getAccountShape } from "../utils";

const sync = makeSync({ getAccountShape });

const receive = makeAccountBridgeReceive();

export const accountBridge: AccountBridge<Transaction> = {
  sync,
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  signOperation,
  broadcast,
  estimateMaxSpendable,
  receive,
};
