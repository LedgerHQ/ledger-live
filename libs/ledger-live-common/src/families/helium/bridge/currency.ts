import { makeScanAccounts } from "../../../bridge/jsHelpers";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import { CurrencyBridge } from "@ledgerhq/types-live";
import { setHeliumPreloadData } from "../preloadedData";
import { getAccountShape } from "../utils";
import { fetchVoteData } from "../api/sdk-getVoteData";
import { HeliumVote } from "../types";

const scanAccounts = makeScanAccounts({ getAccountShape });

export const currencyBridge: CurrencyBridge = {
  preload: async (currency: CryptoCurrency) => {
    const votes = await fetchVoteData(currency);
    setHeliumPreloadData({
      votes,
    });
    return Promise.resolve({
      votes,
    });
  },
  hydrate: (data: { votes?: HeliumVote[] }) => {
    if (!data || typeof data !== "object") return;
    const { votes } = data;

    setHeliumPreloadData({
      votes: votes ?? [],
    });
  },
  scanAccounts,
};
