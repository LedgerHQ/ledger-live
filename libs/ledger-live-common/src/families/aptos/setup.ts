// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
