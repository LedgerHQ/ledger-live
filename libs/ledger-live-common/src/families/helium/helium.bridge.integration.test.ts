import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { DatasetTest } from "@ledgerhq/types-live";
import { BigNumber } from "bignumber.js";
import {
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
} from "@ledgerhq/errors";
import type { Transaction } from "./types";

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    helium: {
      scanAccounts: [
        {
          name: "helium seed 1",
          apdus: `
          => e00200000d038000002c8000038880000000
          <= 000141fd43c78e228b8995002cf92cb40749cdc284d9cd14eba86ac26f8688333cc205b19d0b9000
          => e002000015058000002c80000388800000008000000080000000
          <= 000141fd43c78e228b8995002cf92cb40749cdc284d9cd14eba86ac26f8688333cc205b19d0b9000
          => e002000115058000002c80000388800000018000000080000000
          <= 0001b0e74eb51de315e201a8bc9a9521dc7fb1c13545a8edecbda113b0db6207fd757cac4c4b9000
          `,
        },
      ],
      FIXME_ignorePreloadFields: ["votes"],
      accounts: [
        {
          FIXME_tests: ["balance is sum of ops"],
          raw: {
            id: `js:2:helium:13DTKSEQYRfMgUHXFzgWqUKqokcpd2E4XBF9whPWXJZmT8iBw6N:helium`,
            seedIdentifier: `23ab82f57f7ab4f19097c1a118df91afc5077954d6a012e0086a418ba1559038`,
            name: "Helium 1",
            derivationMode: "helium",
            index: 0,
            freshAddress: "13DTKSEQYRfMgUHXFzgWqUKqokcpd2E4XBF9whPWXJZmT8iBw6N",
            freshAddressPath: "44'/904'/0'/0'/0'",
            freshAddresses: [
              {
                address: "13DTKSEQYRfMgUHXFzgWqUKqokcpd2E4XBF9whPWXJZmT8iBw6N",
                derivationPath: "44'/904'/0'/0'/0'",
              },
            ],
            blockHeight: 0,
            operations: [],
            pendingOperations: [],
            currencyId: "helium",
            unitMagnitude: 8,
            lastSyncDate: "",
            balance: "102815201",
          },
          transactions: [
            // HERE WE WILL INSERT OUR test
            {
              name: "Same as Recipient",
              transaction: (t) => ({
                ...t,
                amount: new BigNumber(100),
                recipient:
                  "13DTKSEQYRfMgUHXFzgWqUKqokcpd2E4XBF9whPWXJZmT8iBw6N",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {
                  recipient: new InvalidAddressBecauseDestinationIsAlsoSource(),
                },
                warnings: {},
              },
            },
            {
              name: "send",
              transaction: (t) => ({
                ...t,
                amount: new BigNumber("1000"),
                recipient:
                  "1br5rK7f4MCpHuezwJjxywLr4pmeMoRMtrkfurNfXAAtJ2Xuk8v",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {},
                warnings: {},
              },
            },
            {
              name: "send amount more than fees + base reserve",
              transaction: (t, account) => ({
                ...t,
                amount: account.balance.plus(1000),
                recipient:
                  "1br5rK7f4MCpHuezwJjxywLr4pmeMoRMtrkfurNfXAAtJ2Xuk8v",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
              },
            },
            {
              name: "send more than base reserve",
              transaction: (t, account) => ({
                ...t,
                amount: account.balance.plus("100"),
                recipient:
                  "1br5rK7f4MCpHuezwJjxywLr4pmeMoRMtrkfurNfXAAtJ2Xuk8v",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
              },
            },
            {
              name: "send Token",
              transaction: (t) => ({
                ...t,
                amount: new BigNumber("1000"),
                recipient:
                  "1br5rK7f4MCpHuezwJjxywLr4pmeMoRMtrkfurNfXAAtJ2Xuk8v",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {},
                warnings: {},
                amount: new BigNumber("1000"),
              },
            },
            {
              name: "send Token - more than available",
              transaction: (t) => ({
                ...t,
                amount: new BigNumber("999999999999999999"),
                recipient:
                  "1br5rK7f4MCpHuezwJjxywLr4pmeMoRMtrkfurNfXAAtJ2Xuk8v",
                model: {
                  mode: "send",
                  memo: "",
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
              },
            },
          ],
        },
      ],
    },
  },
};

testBridge(dataset);
