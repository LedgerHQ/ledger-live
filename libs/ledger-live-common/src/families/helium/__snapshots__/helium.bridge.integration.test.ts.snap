// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helium currency bridge scanAccounts helium seed 1 1`] = `
Array [
  Object {
    "balance": "8000000",
    "currencyId": "helium",
    "derivationMode": "helium",
    "freshAddress": "13SonXS6dNknZvScmdJDL2oAhL3UdTxSXKnYTDce54LnPZEjUXL",
    "freshAddressPath": "44'/904'/0'/0'/0'",
    "freshAddresses": Array [
      Object {
        "address": "13SonXS6dNknZvScmdJDL2oAhL3UdTxSXKnYTDce54LnPZEjUXL",
        "derivationPath": "44'/904'/0'/0'/0'",
      },
    ],
    "heliumResources": Object {
      "stakes": "[]",
    },
    "id": "js:2:helium:13SonXS6dNknZvScmdJDL2oAhL3UdTxSXKnYTDce54LnPZEjUXL:helium",
    "index": 0,
    "name": "Helium 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "41fd43c78e228b8995002cf92cb40749cdc284d9cd14eba86ac26f8688333cc2",
    "spendableBalance": "8000000",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 8,
    "used": true,
  },
  Object {
    "balance": "0",
    "currencyId": "helium",
    "derivationMode": "helium",
    "freshAddress": "14HewXGYzupa96Rm7KBvir8ur6QrKXN4h2ykFCMxtj3aUAb4wvz",
    "freshAddressPath": "44'/904'/1'/0'/0'",
    "freshAddresses": Array [
      Object {
        "address": "14HewXGYzupa96Rm7KBvir8ur6QrKXN4h2ykFCMxtj3aUAb4wvz",
        "derivationPath": "44'/904'/1'/0'/0'",
      },
    ],
    "heliumResources": Object {
      "stakes": "[]",
    },
    "id": "js:2:helium:14HewXGYzupa96Rm7KBvir8ur6QrKXN4h2ykFCMxtj3aUAb4wvz:helium",
    "index": 1,
    "name": "Helium 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "41fd43c78e228b8995002cf92cb40749cdc284d9cd14eba86ac26f8688333cc2",
    "spendableBalance": "0",
    "starred": false,
    "subAccounts": Array [],
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 8,
    "used": false,
  },
]
`;

exports[`helium currency bridge scanAccounts helium seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "13SonXS6dNknZvScmdJDL2oAhL3UdTxSXKnYTDce54LnPZEjUXL",
      "blockHash": "1500661",
      "blockHeight": 1500661,
      "contract": undefined,
      "extra": Object {},
      "fee": "35000",
      "hash": "Ru3NfikAYZX-QmA-TpBYjHrGtLjkvZFvl2c2FYKD49M",
      "id": "Ru3NfikAYZX-QmA-TpBYjHrGtLjkvZFvl2c2FYKD49M",
      "operator": undefined,
      "recipients": Array [
        "13SonXS6dNknZvScmdJDL2oAhL3UdTxSXKnYTDce54LnPZEjUXL",
      ],
      "senders": Array [
        "14s1pLdZ3Z8EobTURy7Jn1hBdJDdNyjYZhNvrssNXbBRPZ2PaHn",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "transactionSequenceNumber": 38,
      "type": "IN",
      "value": "8000000",
    },
  ],
  Array [],
]
`;
